pkg/server/plugin/upstreamauthority/awspca/pca_client.go

```diff
@@ -1,49 +1,67 @@
-	GetCertificateWithContext(aws.Context, *acmpca.GetCertificateInput, ...request.Option) (*acmpca.GetCertificateOutput, error)
+	DescribeCertificateAuthority(context.Context, *acmpca.DescribeCertificateAuthorityInput, ...func(*acmpca.Options)) (*acmpca.DescribeCertificateAuthorityOutput, error)
+	IssueCertificate(context.Context, *acmpca.IssueCertificateInput, ...func(*acmpca.Options)) (*acmpca.IssueCertificateOutput, error)
+	GetCertificate(context.Context, *acmpca.GetCertificateInput, ...func(*acmpca.Options)) (*acmpca.GetCertificateOutput, error)
 }
 
-func newPCAClient(config *Configuration) (PCAClient, error) {
-	awsConfig := &aws.Config{
-		Region:   aws.String(config.Region),
-		Endpoint: aws.String(config.Endpoint),
+type certificateIssuedWaiter interface {
+	Wait(context.Context, *acmpca.GetCertificateInput, time.Duration, ...func(*acmpca.CertificateIssuedWaiterOptions)) error
+	WaitForOutput(context.Context, *acmpca.GetCertificateInput, time.Duration, ...func(*acmpca.CertificateIssuedWaiterOptions)) (*acmpca.GetCertificateOutput, error)
+}
+
+func newPCAClient(ctx context.Context, cfg *Configuration) (PCAClient, error) {
+	var endpointResolver aws.EndpointResolverWithOptions
+	if cfg.Endpoint != "" {
+		endpointResolver = aws.EndpointResolverWithOptionsFunc(func(service, region string, options ...interface{}) (aws.Endpoint, error) {
+			if service == acmpca.ServiceID && region == cfg.Region {
+				return aws.Endpoint{
+					PartitionID:   "aws",
+					URL:           cfg.Endpoint,
+					SigningRegion: region,
+				}, nil
+			}
+
+			return aws.Endpoint{}, fmt.Errorf("unknown endpoint requested")
+		})
 	}
 
-	// Optional: Assuming role
-	if config.AssumeRoleARN != "" {
-		staticsess, err := session.NewSession(&aws.Config{Credentials: awsConfig.Credentials})
+	var credsProvider aws.CredentialsProvider
+	switch {
+	case cfg.AssumeRoleARN != "":
+		stsClient := sts.NewFromConfig(aws.Config{})
+		credsProvider = stscreds.NewAssumeRoleProvider(stsClient, cfg.AssumeRoleARN)
+	default:
+		awsCfg, err := config.LoadDefaultConfig(ctx, config.WithRegion(cfg.Region), config.WithEndpointResolverWithOptions(endpointResolver))
 		if err != nil {
 			return nil, err
 		}
-		awsConfig.Credentials = credentials.NewCredentials(&stscreds.AssumeRoleProvider{
-			Client:   sts.New(staticsess),
-			RoleARN:  config.AssumeRoleARN,
-			Duration: 15 * time.Minute,
-		})
-	}
 
-	awsSession, err := session.NewSession(awsConfig)
-	if err != nil {
-		return nil, err
+		credsProvider = awsCfg.Credentials
 	}
 
-	return acmpca.New(awsSession), nil
+	return acmpca.NewFromConfig(aws.Config{
+		Region:                      cfg.Region,
+		EndpointResolverWithOptions: endpointResolver,
+		Credentials:                 credsProvider,
+	}), nil
+}
+
+func newCertificateIssuedWaiter(client acmpca.GetCertificateAPIClient, optFns ...func(*acmpca.CertificateIssuedWaiterOptions)) certificateIssuedWaiter {
+	return acmpca.NewCertificateIssuedWaiter(client, optFns...)
```

Comment:
I wonder if we shouldn't set our own custom Retryable option until https://github.com/aws/aws-sdk-go-v2/issues/1585 is fixed...

Reply:
Maybe? It is also something that we need to maintain over time though. I think I'd rather see the fix go into the AWS SDK and pick it up from there so we don't have a Retryable that gets stale and out of sync with AWS API call patterns.

I'm trying to think about the worst that will happen with the existing Retryable behavior. It seems like it will just take longer to fail an operation than is necessary? With an upstream CA, rotation windows should in practice be much larger than this retry window, so we might be technically ok.

Reply:
It would probably be good to get confirmation from the AWS SDK maintainers that the behavior is actually a bug as well. It seems like it is, but I'm not so familiar with these APIs.